ddtrace/internal/datadog/profiling/src/exporter.cpp

```diff
@@ -0,0 +1,560 @@
+        get_value_idx("lock-release-hold", "nanoseconds");
+    val_idx.lock_release_count = get_value_idx("lock-release", "count");
+  }
+  if (type_mask & ProfileType::Allocation) {
+    val_idx.alloc_space = get_value_idx("alloc-space", "bytes");
+    val_idx.alloc_count = get_value_idx("alloc-samples", "count");
+  }
+  if (type_mask & ProfileType::Heap) {
+    val_idx.heap_space = get_value_idx("heap-space", "bytes");
+  }
+
+  values.resize(samplers.size());
+  std::fill(values.begin(), values.end(), 0);
+
+  ddog_prof_Period default_sampler = {
+      samplers[0], 1}; // Mandated by pprof, but probably unused
+  ddog_profile = ddog_prof_Profile_new({&samplers[0], samplers.size()},
+                                       &default_sampler, nullptr);
+
+  // Prepare for use
+  reset();
+
+  // Initialize the size for buffers
+  locations.reserve(2040);
+  lines.reserve(2048);
+  strings.reserve(8192);
+}
+
+Profile::~Profile() { ddog_prof_Profile_drop(ddog_profile); }
+
+bool Profile::reset() {
+  if (!ddog_prof_Profile_reset(ddog_profile, nullptr)) {
+    errmsg = "Unable to reset profile";
+    return false;
+  }
+  return true;
+}
+
+bool Profile::start_sample(unsigned int nframes) {
+  strings.clear();
+  clear_buffers();
+  this->nframes = nframes;
+  return true;
+}
+
+void Profile::push_frame_impl(std::string_view name, std::string_view filename,
+                              uint64_t address, int64_t line) {
+
+  // Ensure strings are stored.
+  // Slightly wasteful since it requires allocating another string.
+  auto insert_or_get = [&](std::string_view sv) -> std::string_view {
+    std::string str(sv);
+    auto [it, _] = strings.insert(std::move(str));
+    return *it;
+  };
+  name = insert_or_get(name);
+  filename = insert_or_get(filename);
+
+  lines.push_back({
+      .function =
+          {
+              .name = to_slice(name),
+              .system_name = {},
+              .filename = to_slice(filename),
+              .start_line = 0,
+          },
+      .line = line,
+  });
+
+  locations.push_back({
+      {},
+      address,
+      {&lines.back(), 1},
+      false,
+  });
+}
+
+void Profile::push_frame(std::string_view name, std::string_view filename,
+                         uint64_t address, int64_t line) {
+
+  push_frame_impl(name, filename, address, line);
+}
+
+void Profile::push_label(const ExportLabelKey key, std::string_view val) {
+  // libdatadog checks the labels when they get flushed, which slightly
+  // de-localizes the error message.  Roll with it for now.
+  constexpr std::array<std::string_view,
+                       static_cast<size_t>(ExportLabelKey::_Length)>
+      keys = {EXPORTER_LABELS(X_STR)};
+  std::string_view key_sv = keys[static_cast<size_t>(key)];
+  labels[cur_label].key = to_slice(key_sv);
+
+  // Label may not persist, so it needs to be saved
+  auto [it, _] = strings.insert(std::string{val});
+  labels[cur_label].str = to_slice(*it);
+  cur_label++;
```

Comment:
It is probably best to add some safety bound check around this.

Reply:
Ah, good point.  The label definition was wrong and should have been set to the converted `_Length` of the enum.  Fixed in 61f10cc50